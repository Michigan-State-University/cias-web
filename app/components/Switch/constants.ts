export enum LabelPosition {
  Left = 'left',
  Right = 'right',
}
