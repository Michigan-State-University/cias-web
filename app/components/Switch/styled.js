import styled from 'styled-components';
import { themeColors, colors } from 'theme';
import { margin, padding } from '../BaseComponentStyles';

const SwitchWrapper = styled.label`
  position: relative;
  display: inline-block;
  width: 40px;
  height: 21px;
  ${margin};
  ${padding};
`;

const SwitchInput = styled.input.attrs({ type: 'checkbox' })`
  opacity: 0;
  width: 0;
  height: 0;
`;

const Slider = styled.span`
  position: absolute;
  top: 0;
  left: 0;
  right: 0;
  bottom: 0;

  transition: 0.4s;

  cursor: pointer;
  background-color: ${colors.blueHaze};
  border-radius: 21px;

  ${SwitchInput}:checked + & {
    background-color: ${themeColors.secondary};
  }

  ${SwitchInput}:focus + & {
    box-shadow: 0 0 1px ${themeColors.secondary};
  }

  &:before {
    position: absolute;
    content: '';
    left: 4px;
    bottom: 4px;

    transition: 0.4s;

    height: 13px;
    width: 13px;
    background-color: white;
    border-radius: 50%;

    ${SwitchInput}:checked + & {
      transform: translateX(18px);
    }
  }
`;

export { SwitchWrapper, SwitchInput, Slider };
