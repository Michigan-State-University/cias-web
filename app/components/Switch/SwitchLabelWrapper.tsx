import React, { memo, ReactNode } from 'react';

import Row from 'components/Row';

import { LabelPosition } from './constants';

type Props = {
  children: ReactNode;
  labelPosition: LabelPosition;
};

const SwitchLabelWrapper = ({
  children,
  labelPosition = LabelPosition.Left,
}: Props): JSX.Element => {
  const [Switch, Label] = children as JSX.Element[];

  if (!Label) return Switch;

  return (
    <Row justify="between" align="center" width="100%">
      {labelPosition === LabelPosition.Left && Label}

      {Switch}

      {labelPosition === LabelPosition.Right && Label}
    </Row>
  );
};

export default memo(SwitchLabelWrapper);
