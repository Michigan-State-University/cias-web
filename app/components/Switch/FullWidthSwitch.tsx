import { ComponentProps, FunctionComponent } from 'react';
// @ts-ignore
import styled from 'styled-components';

import Switch from './Switch';

const FullWidthSwitch: FunctionComponent<
  ComponentProps<typeof Switch>
  // @ts-ignore
> = styled(Switch)`
  width: 100% !important;
`;

export default FullWidthSwitch;
