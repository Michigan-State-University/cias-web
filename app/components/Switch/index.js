import React from 'react';
import PropTypes from 'prop-types';
import { SwitchWrapper, SwitchInput, Slider } from './styled';

const Switch = props => (
  <SwitchWrapper {...props}>
    {console.log('eeeeee', props.e ? props.checked : '')}
    <SwitchInput
      checked={props.checked}
      onChange={event => props.onToggle(event.target.checked)}
    />
    <Slider />
  </SwitchWrapper>
);

Switch.propTypes = {
  checked: PropTypes.bool,
  onToggle: PropTypes.func,
};

Switch.defaultProps = {
  checked: false,
};

export default Switch;
