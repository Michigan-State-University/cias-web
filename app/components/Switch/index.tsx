import React, { CSSProperties, memo, ReactNode, useCallback } from 'react';

import { Slider, SwitchInput, SwitchWrapper } from './styled';
import { LabelPosition } from './constants';
import SwitchLabelWrapper from './SwitchLabelWrapper';

type Props = {
  checked: boolean;
  children?: ReactNode;
  disabled?: boolean;
  id?: string;
  labelPosition?: LabelPosition;
  onToggle: (value: boolean) => void;
} & Partial<CSSProperties>; // `CSSProperties` for now; after refactor use specific type `SwitchWrapper` can get

const Switch = ({
  checked = false,
  children,
  disabled = false,
  id,
  labelPosition = LabelPosition.Left,
  onToggle,
  ...props
}: Props): JSX.Element => {
  const onClick = useCallback(() => {
    onToggle(!checked);
  }, [checked, onToggle]);

  // needed empty function for HTML correctness
  const onChange = useCallback(() => {}, []);

  return (
    <SwitchLabelWrapper labelPosition={labelPosition}>
      <SwitchWrapper onClick={onClick} {...props}>
        <SwitchInput
          id={id}
          data-cy="branching-intervention-toggle"
          data-testid="switch-input"
          disabled={disabled}
          checked={checked}
          onChange={onChange}
        />
        <Slider />
      </SwitchWrapper>

      {children && <label htmlFor={id}>{children}</label>}
    </SwitchLabelWrapper>
  );
};

export { LabelPosition };
export default memo(Switch);
