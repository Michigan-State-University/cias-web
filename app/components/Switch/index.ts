import Switch, { Props } from './Switch';
import FullWidthSwitch from './FullWidthSwitch';

import { LabelPosition } from './constants';

export { Switch, FullWidthSwitch, LabelPosition };
export type { Props };
export default Switch;
