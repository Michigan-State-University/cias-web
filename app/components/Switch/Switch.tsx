import React, {
  AriaAttributes,
  ChangeEvent,
  ChangeEventHandler,
  memo,
  ReactNode,
} from 'react';

import Row from 'components/Row';
import { LayoutProps, MarginProps } from 'components/BaseComponentStyles';
import Loader from 'components/Loader';

import SwitchLabelWrapper from './SwitchLabelWrapper';

import {
  LabelContent,
  Slider,
  StyledLabel,
  SwitchInput,
  SwitchWrapper,
} from './styled';
import { LabelPosition } from './constants';

type NativeChangeHandlerProps = {
  onToggle?: undefined;
  onChange: ChangeEventHandler<HTMLInputElement>;
  nativeChangeHandler: true;
};

type CustomChangeHandlerProps = {
  onToggle: (value: boolean, event: ChangeEvent<HTMLInputElement>) => void;
  onChange?: undefined;
  nativeChangeHandler?: false;
};

type ChangeHandlerProps = NativeChangeHandlerProps | CustomChangeHandlerProps;

export type Props = {
  checked?: boolean;
  children?: ReactNode;
  className?: string;
  disabled?: boolean;
  id: string;
  labelPosition?: LabelPosition;
  labelOffset?: number;
  onBlur?: HTMLInputElement['onblur'];
  ariaLabel?: AriaAttributes['aria-label'];
  loading?: boolean;
} & ChangeHandlerProps &
  LayoutProps &
  MarginProps;

const Switch = ({
  checked = false,
  children,
  className,
  disabled = false,
  id,
  labelPosition = LabelPosition.Left,
  labelOffset,
  onToggle,
  onChange,
  onBlur,
  nativeChangeHandler,
  ariaLabel,
  loading,
  ...props
}: Props): JSX.Element => {
  const handleOnChange: ChangeEventHandler<HTMLInputElement> = (event) => {
    const {
      target: { checked: newValue },
    } = event;

    return onToggle?.(newValue, event);
  };

  return (
    <Row justify="between" align="center" className={className}>
      <SwitchInput
        id={id}
        data-cy="branching-intervention-toggle"
        data-testid="switch-input"
        aria-label={ariaLabel}
        disabled={disabled}
        checked={checked}
        onChange={nativeChangeHandler ? onChange : handleOnChange}
        onBlur={onBlur}
      />

      <StyledLabel htmlFor={id}>
        <SwitchLabelWrapper labelPosition={labelPosition} {...props}>
          <SwitchWrapper>
            {!loading && <Slider />}
            {loading && <Loader type="inline" size={21} />}
          </SwitchWrapper>

          <LabelContent
            $labelPosition={labelPosition}
            $labelOffset={labelOffset}
          >
            {children}
          </LabelContent>
        </SwitchLabelWrapper>
      </StyledLabel>
    </Row>
  );
};

export default memo(Switch);
