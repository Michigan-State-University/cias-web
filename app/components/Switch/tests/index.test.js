import React from 'react';
import { render, fireEvent } from 'react-testing-library';
import 'jest-styled-components';

import Switch from '../index';

describe('<Switch />', () => {
  const defaultProps = {
    onToggle: jest.fn(),
  };
  it('Expect to not log errors in console', () => {
    const spy = jest.spyOn(global.console, 'error');
    render(<Switch {...defaultProps} />);
    expect(spy).not.toHaveBeenCalled();
  });
  it('should match the snapshot', () => {
    const { container } = render(<Switch {...defaultProps} />);
    expect(container).toMatchSnapshot();
  });
  it('should call on toggle', () => {
    const { getByTestId } = render(<Switch {...defaultProps} />);
    const switchInput = getByTestId('switch-input');
    fireEvent.click(switchInput, { target: { checked: true } });
    expect(defaultProps.onToggle).toHaveBeenCalled();
  });
});
