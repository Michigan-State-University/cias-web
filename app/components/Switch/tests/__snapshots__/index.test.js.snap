// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Switch /> should match the snapshot 1`] = `
.c0 {
  position: relative;
  display: inline-block;
  min-width: 40px;
  height: 21px;
}

.c2 {
  opacity: 0;
  width: 0;
  height: 0;
}

.c4 {
  position: absolute;
  top: 0;
  left: 0;
  right: 0;
  bottom: 0;
  -webkit-transition: 0.4s;
  transition: 0.4s;
  cursor: pointer;
  background-color: #CED3E4;
  border-radius: 21px;
}

.c1:checked + .c3 {
  background-color: #C866EA;
}

.c1:focus + .c3 {
  box-shadow: 0 0 1px #C866EA;
}

.c1:disabled + .c3 {
  opacity: 0.5;
  cursor: not-allowed;
}

.c4:before {
  position: absolute;
  content: '';
  left: 4px;
  bottom: 4px;
  -webkit-transition: 0.4s;
  transition: 0.4s;
  height: 13px;
  width: 13px;
  background-color: white;
  border-radius: 50%;
}

.c1:checked + .c3:before {
  -webkit-transform: translateX(18px);
  -ms-transform: translateX(18px);
  transform: translateX(18px);
}

<div>
  <label
    class="c0"
  >
    <input
      class="c1 c2"
      data-cy="branching-intervention-toggle"
      data-testid="switch-input"
      type="checkbox"
    />
    <span
      class="c3 c4"
    />
  </label>
</div>
`;
